keanu-project/src/test/java/io/improbable/keanu/algorithms/graphtraversal/DifferentiableCheckerTest.java

```diff
@@ -0,0 +1,166 @@
+        assertEquals(true, differentiable);
+    }
+
+    @Test
+    public void graphWithAssertIsDiffable() {
+        GaussianVertex gaussianA = new GaussianVertex(5., 1.);
+        gaussianA.lessThan(new ConstantDoubleVertex(90)).assertTrue();
+        GaussianVertex gaussianB = new GaussianVertex(gaussianA, 1.);
+        BayesianNetwork bayesianNetwork = new BayesianNetwork(gaussianB.getConnectedGraph());
+        boolean differentiable = DifferentiableChecker.isDifferentiable(bayesianNetwork.getLatentOrObservedVertices());
+        assertEquals(true, differentiable);
+    }
+
+    @Test
+    public void multipleNonDiffableWithConstantParents() {
+        GaussianVertex gaussianA = new GaussianVertex(5., 1.);
+        GaussianVertex gaussianB = new GaussianVertex(gaussianA, 1.);
+        gaussianA.observe(4.);
+        gaussianB.observe(4.);
+
+        DoubleVertex resultVertex = gaussianA.multiply(gaussianB).plus(gaussianB);
+        DoubleVertex nonDiffableA = new FloorVertex(resultVertex);
+        DoubleVertex nonDiffableB = new FloorVertex(resultVertex);
+        DoubleVertex nonDiffableC = new FloorVertex(resultVertex);
+        DoubleVertex nonDiffableD = new FloorVertex(resultVertex);
+        DoubleVertex nonDiffableE = new FloorVertex(resultVertex);
+        DoubleVertex nonDiffSum = nonDiffableA.plus(nonDiffableB).plus(nonDiffableC).plus(nonDiffableD).plus(nonDiffableE);
+        GaussianVertex gaussianSum = new GaussianVertex(nonDiffSum, 1.);
+        BayesianNetwork bayesianNetwork = new BayesianNetwork(gaussianSum.getConnectedGraph());
+        boolean differentiable = DifferentiableChecker.isDifferentiable(bayesianNetwork.getLatentOrObservedVertices());
+        assertEquals(true, differentiable);
+    }
+
+    @Test
+    public void observedVertexWithNonDiffableParentIsntDiffable() {
+        GaussianVertex gaussianA = new GaussianVertex(5., 1.);
+        FloorVertex nonDiffable = new FloorVertex(gaussianA);
+        GaussianVertex observed = new GaussianVertex(nonDiffable, 1.);
+        observed.observe(4.);
+        BayesianNetwork bayesianNetwork = new BayesianNetwork(observed.getConnectedGraph());
+        boolean differentiable = DifferentiableChecker.isDifferentiable(bayesianNetwork.getLatentOrObservedVertices());
+        assertEquals(false, differentiable);
+    }
+
+    @Test
+    public void observedVertexWithNonDiffableConstantParentIsDiffable() {
+        DoubleVertex constDouble = new ConstantDoubleVertex(4);
+        FloorVertex nonDiffable = new FloorVertex(constDouble);
+        GaussianVertex observed = new GaussianVertex(nonDiffable, 1.);
+        observed.observe(4.);
+        BayesianNetwork bayesianNetwork = new BayesianNetwork(observed.getConnectedGraph());
+        boolean differentiable = DifferentiableChecker.isDifferentiable(bayesianNetwork.getLatentOrObservedVertices());
+        assertEquals(true, differentiable);
+    }
+
```

Comment:
I have added a mock to ensure that the cache is working by mocking the `isObserved` on a vertex. It would be good to add the check you mention, however as `isDifferentiable()` is final I can't mock it or spy on it.